File: CLEO/CarGeneratorWPL.js
/// <reference path=".config/gta_iv.d.ts" />
/// <reference no-default-lib="true" />

var PlayerChar = new Player(0).getChar();

const SaveKey = 89; // Y
const GeneratorFlags = [1632, 1633, 1760, 1888, 1889];

function SaveCarGeneratorToLogFile() {
    if (!Char.DoesExist(PlayerChar)) {
        Text.PrintStringWithLiteralStringNow("string", "CarGeneratorWPL: ~r~Can't save at the moment", 4000, true);
        return;
    }

    if (!PlayerChar.isInAnyCar()) {
        return;
    }

    let playerCar = PlayerChar.getCarIsUsing();

    let carPosX = playerCar.getCoordinates().x;
    let carPosY = playerCar.getCoordinates().y;
    let carPosZ = playerCar.getCoordinates().z;
    let carHeading = playerCar.getHeading();

    let headingInDegrees = carHeading * (Math.PI / 180.0);
    let forwardX = -Math.sin(headingInDegrees) * 7.0;
    let forwardY = Math.cos(headingInDegrees) * 7.0;

    let currentFlag = GeneratorFlags[Math.RandomIntInRange(0, GeneratorFlags.length)];

    let line1 = `${carPosX}, ${carPosY}, ${carPosZ}, ${forwardX}, ${forwardY}, `;
    let line2 = `3, hash:0, -1, -1, -1, -1, ${currentFlag}, 0, 0\n`;
    
    let file = File.Open("CLEO/CarGeneratorWPL.log");
    if (file == -1) {
        Text.PrintStringWithLiteralStringNow("string", "CarGeneratorWPL: ~r~Failed to save the car generator!", 4000, true);
    } else {
        file.writeString(line1);
        file.writeString(line2);
        file.close();

        Text.PrintStringWithLiteralStringNow("string", "CarGeneratorWPL: The car generator is saved", 3000, true);
    }
}

while(true) {
    wait(0);

    if (Pad.IsKeyDown(SaveKey)) {
        SaveCarGeneratorToLogFile();
    }
}

File: IVSDKDotNet/CarGeneratorWPL.cs
using System;
using System.Numerics;
using System.Windows.Forms;
using System.IO;

using IVSDKDotNet;
using static IVSDKDotNet.Native.Natives;

namespace CarGeneratorWPL
{
    public class CarGeneratorWPL : Script
    {
        readonly int[] GeneratorFlags = { 1632, 1633, 1760, 1888, 1889, };

        Random RandomFlag = new Random();

        public CarGeneratorWPL()
        {
            KeyDown += OnKeyDown;
        }

        private void OnKeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode != Keys.Y)
            {
                return;
            }

            SaveCarGeneratorToLogFile();
        }

        private void SaveCarGeneratorToLogFile()
        {
            int PlayerIndex = CONVERT_INT_TO_PLAYERINDEX(GET_PLAYER_ID());

            int PlayerChar;
            GET_PLAYER_CHAR(PlayerIndex, out PlayerChar);

            if (!DOES_CHAR_EXIST(PlayerChar))
            {
                PRINT_STRING_WITH_LITERAL_STRING_NOW("string", "CarGeneratorWPL: ~r~Can't save at the moment", 4000, true);
                return;
            }

            if (!IS_CHAR_IN_ANY_CAR(PlayerChar))
            {
                return;
            }

            int PlayerVehicle;
            GET_CAR_CHAR_IS_USING(PlayerChar, out PlayerVehicle);

            Vector3 VehiclePosition;
            GET_CAR_COORDINATES(PlayerVehicle, out VehiclePosition);

            float VehicleHeading;
            GET_CAR_HEADING(PlayerVehicle, out VehicleHeading);

            double HeadingInDegrees = VehicleHeading * (Math.PI / 180.0);
            double ForwardX = -Math.Sin(HeadingInDegrees) * 7.0;
            double ForwardY = Math.Cos(HeadingInDegrees) * 7.0;

            int CurrentFlag = RandomFlag.Next(0, GeneratorFlags.Length);

            string[] Line = { string.Format("{0}, {1}, {2}, {3:f3}, {4:f3}, 3, hash:0, -1, -1, -1, -1, {5}, 0, 0",
                                            VehiclePosition.X, VehiclePosition.Y, VehiclePosition.Z, ForwardX, ForwardY, GeneratorFlags[CurrentFlag]) };

            File.AppendAllLines(Path.Combine("IVSDKDotNet\\scripts\\", "CarGeneratorWPL.ivsdk.log"), Line);

            PRINT_STRING_WITH_LITERAL_STRING_NOW("string", "CarGeneratorWPL: The car generator is saved", 3000, true);
        }
    }
}
